### ISSUE TITLE
[Results] TensorFlow Lite v2.17.0 cross-compiled for riscv64 — libtensorflow-lite.a (21MB, 243 object files, elf64-littleriscv)

### ISSUE THREAD

**Issue by trg-rgb** [clusterchallenge/Hardware-Abstraction-Layer-Transitional-Libraries#17]
**Applicant:** Tanmay Gulhane · MIT World Peace University, Pune
**Project:** Broadening the RISC-V High Precision Code Base and Reach
**Mentor:** Kurt Keville (MIT)
**Work Repository:** [github.com/trg-rgb/riscv-hpc-port](https://github.com/trg-rgb/riscv-hpc-port)
 
---
 
## What This Issue Documents
 
A complete cross-compilation of TensorFlow Lite v2.17.0 for `riscv64`, producing a verified `elf64-littleriscv` static library. This is the first ML inference runtime port attempted in this applicant pool, and targets the Community AI/ML codes listed in the program's 400-code sheet directly.
 
I built and deployed a 6-class CNN for groundnut leaf disease detection (Gold Medal, Sci Quest 2025, MIT-WPU) — I understand what TF Lite is for, not just how to compile it. The goal during the mentorship is to run real model inference on riscv64, not just produce a library binary.
 
> 🤗 Deployed model: [huggingface.co/spaces/tanmaytrg/Pretrained_CNN_6_class_image_classification](https://huggingface.co/spaces/tanmaytrg/Pretrained_CNN_6_class_image_classification)
 
---
 
## Environment
 
| Component | Value |
|---|---|
| TensorFlow Lite version | `v2.17.0` (tag `ad6d8cc177d`) |
| C/C++ cross-compiler | `riscv64-linux-gnu-gcc 15.2.0` |
| Target architecture | `RV64GC` |
| Binary format | `elf64-littleriscv` |
| Host | Ubuntu WSL2 on x86_64 |
| Build system | CMake 4.2 |
 
---
 
## Build Configuration
 
```cmake
cmake ~/tensorflow-v2.17.0/tensorflow/lite \
  -DCMAKE_TOOLCHAIN_FILE=riscv64-toolchain.cmake \
  -DCMAKE_BUILD_TYPE=Release \
  -DTFLITE_ENABLE_XNNPACK=OFF \
  -DTFLITE_ENABLE_RUY=OFF \
  -DBUILD_SHARED_LIBS=OFF
```
 
### Flag Rationale
 
| Flag | Reason |
|---|---|
| `-DTFLITE_ENABLE_XNNPACK=OFF` | XNNPACK's core kernels use NEON (ARM) and SSE2 (x86) intrinsics. On base RV64GC without the V extension these won't compile. Scalar baseline first — RVV paths are a follow-on task. |
| `-DTFLITE_ENABLE_RUY=OFF` | Ruy has architecture-specific fast paths not present on riscv64. Same rationale as XNNPACK. |
| `-DBUILD_SHARED_LIBS=OFF` | Static library is self-contained and portable under QEMU without dynamic linker path issues. |
 
---
 
## Result
 
```
[100%] Linking CXX static library libtensorflow-lite.a
[100%] Built target tensorflow-lite
```
 
| Metric | Value |
|---|---|
| Output | `libtensorflow-lite.a` |
| Size | 21 MB |
| Object files in archive | 243 |
| Binary format | `elf64-littleriscv` |
| Architecture | `riscv:rv64` |
 
### Verified with `riscv64-linux-gnu-objdump`
 
```
In archive libtensorflow-lite.a:
 
delegate_registry.cc.o:     file format elf64-littleriscv
architecture: riscv:rv64, flags 0x00000011:
HAS_RELOC, HAS_SYMS
 
nnapi_plugin.cc.o:     file format elf64-littleriscv
architecture: riscv:rv64, flags 0x00000011:
HAS_RELOC, HAS_SYMS
 
stable_delegate_registry.cc.o:     file format elf64-littleriscv
architecture: riscv:rv64, flags 0x00000011:
HAS_RELOC, HAS_SYMS
```
 
### Exported TF Lite Symbols (sample from `riscv64-linux-gnu-nm`)
 
These are real TF Lite inference engine symbols compiled for riscv64 — not stubs:
 
```
T tflite::delegates::DelegatePluginRegistry::CreateImpl(...)
T tflite::delegates::StableDelegateRegistry::RegisterStableDelegate(...)
T tflite::ErrorReporter::Report(...)
T tflite::ParseEqual(...)
T tflite::ParseFloor(...)
T tflite::ParsePadV2(...)
T tflite::ParsePrelu(...)
T tflite::ParseRelu6(...)
T tflite::ParseRound(...)
T tflite::ParseRsqrt(...)
```
 
Full library: [`tflite/results/libtensorflow-lite.a`](tflite/results/libtensorflow-lite.a)
Build log: [`tflite/results/tflite_build_results.txt`](tflite/results/tflite_build_results.txt)
 
---
 
### Blocker 1 — Dynamic linker path for QEMU user-mode emulation
 
**Encountered during:** Eigen 5.0.1 test run (same toolchain)
 
Dynamically linked riscv64 binaries fail under `qemu-riscv64` without pointing the emulator at the sysroot:
 
```
qemu-riscv64: Could not open '/lib/ld-linux-riscv64-lp64d.so.1': No such file or directory
```
 
**Fix:**
```bash
QEMU_LD_PREFIX=/usr/riscv64-linux-gnu qemu-riscv64 ./binary
```
 
This is a standard WSL2 cross-compilation gotcha — the riscv64 sysroot is at `/usr/riscv64-linux-gnu`, not `/lib`. Once `QEMU_LD_PREFIX` is set, all dynamically linked riscv64 binaries run correctly. Applied this fix across all Eigen tests — 42/42 passed after correction.
 
### Blocker 2 — `protoc` cross-compiled for riscv64 instead of host x86
 
**Encountered during:** First TF Lite build attempt
 
CMake tried to cross-compile the protobuf compiler (`protoc`) for riscv64 when it needs to run on the host x86 to generate C++ files at build time. Build failed at the protobuf stage:
 
```
make[1]: *** [_deps/protobuf-build/CMakeFiles/protoc.dir/all] Error 2
```
 
**Fix:**
```bash
sudo apt install protobuf-compiler   # install native x86 protoc
rm -rf tflite-build-riscv/*          # clear stale CMake cache
# reconfigure — cmake picks up the native protoc automatically
```
 
This is a known cross-compilation pitfall with any project that uses protobuf code generation. The fix is documented here for reproducibility.
 
---
 
## What's Next
 
`libtensorflow-lite.a` is the static library. The next step is linking it into a working inference binary and running an actual `.tflite` model under `qemu-riscv64`. The target is my own trained groundnut disease CNN exported to TFLite format — end-to-end ML inference on emulated RISC-V, which directly exercises the program's Community AI/ML code targets.
 
Follow-on tasks in roughly this order:
- Build `benchmark_model` binary against `libtensorflow-lite.a` for riscv64
- Export groundnut CNN to `.tflite` flatbuffer
- Run inference under `qemu-riscv64`, verify output matches x86 reference
- Investigate RVV kernel paths — whether XNNPACK or Ruy can be re-enabled with scalar fallback shim (similar approach to Aayush's SSE2 shim)
---
 
## Links
 
- Work repo: [github.com/trg-rgb/riscv-hpc-port](https://github.com/trg-rgb/riscv-hpc-port)
- TF Lite library: [tflite/results/libtensorflow-lite.a](https://github.com/trg-rgb/riscv-hpc-port/tree/main/tflite/results)
- Eigen results (42/42): [eigen/results/](https://github.com/trg-rgb/riscv-hpc-port/tree/main/eigen/results)
- OpenBLAS results: [openblas/results/](https://github.com/trg-rgb/riscv-hpc-port/tree/main/openblas/results)
- Deployed CNN: [huggingface.co/spaces/tanmaytrg/Pretrained_CNN_6_class_image_classification](https://huggingface.co/spaces/tanmaytrg/Pretrained_CNN_6_class_image_classification)
- Contact: tanmaygulhane12@gmail.com

**Comment by Vaibhav805** [comment]
Great work on the TF Lite port, @trg-rgb. I noticed you disabled XNNPACK and Ruy to bypass compilation blockers, forcing TF Lite onto generic C++ reference kernels. While great for a baseline, this introduces massive latency for actual HPC deployments alternative could be Instead of keeping the execution purely scalar, TF Lite can route tensor operations through an external BLAS engine. Since we now have a verified OpenBLAS (rv64gcv) build with RVV 1.0 vector kernels and confirmed numerical parity , linking your framework against it would unlock true hardware acceleration.

**Comment by trg-rgb** [comment]
Thanks @Vaibhav805 , BLAS-backed TF Lite kernels is a valid direction worth exploring. A couple of technical clarifications:

The XNNPACK/Ruy flags weren't workarounds for blockers. Both libraries have no riscv64 fast path at all. Disabling them routes TF Lite to its own optimised reference kernels, which is the correct baseline for a new architecture port.

On the OpenBLAS point: your build used `TARGET=RISCV64_GENERIC`, which is the scalar fallback, same as mine. That's not an RVV build. A genuine RVV-accelerated OpenBLAS requires `-march=rv64gcv` and a vector-capable target, plus `qemu-riscv64 -cpu rv64,v=true` to execute. Worth attempting, but neither of us has that verified yet.

The BLAS integration path for TF Lite is also non-trivial. It requires building with `-DTFLITE_BLAS_LIBRARY` and validating kernel dispatch, not just linking. Planning to explore this after getting baseline inference running.

**Comment by trg-rgb** [comment]
Update (May 21): End-to-end inference now running on riscv64.

Converted ensemble_model.h5 to groundnut_cnn.tflite (59.9MB, INT8 quantized via tf.lite.Optimize.DEFAULT) and ran it against benchmark_model under qemu-riscv64.

Results:
- Init: 227ms
- First inference: 797,720,987 µs
- Inference avg: 781,353,268 µs
- Memory footprint: 100.7MB

Latency is QEMU user-mode emulation overhead. Every riscv64 instruction is emulated on x86. On real hardware this would be orders of magnitude faster. The result proves the full stack works: libtensorflow-lite.a, benchmark_model, groundnut_cnn.tflite, output on riscv64.

Model: tflite/groundnut_cnn.tflite
Benchmark log: tflite/results/benchmark_results.txt
